spring-kafka/src/main/java/org/springframework/kafka/listener/WakingKafkaConsumerTimingAdjuster.java

```diff
@@ -138,4 +154,18 @@ private void doApplyTimingAdjustment(Consumer<?, ?> consumerForTimingAdjustment,
 					"for TopicPartition " + topicPartition);
 		}
 	}
+
+	private static TaskExecutor createTaskExecutor() {
+		ThreadPoolTaskExecutor taskExecutor = new ThreadPoolTaskExecutor();
```

Comment:
Again: why plain `Executors.newSingleThreadExecutor()` is not enough to have a full default control of this?
This way we don't need that `destroy()` overhead in this bean.

Reply:
Sorry, I had answered this [here](https://github.com/spring-projects/spring-kafka/pull/2227#discussion_r849766492), but I think I marked it as resolved before you had a chance to look at it.

In a shorter version, we currently rely on spawning threads to wake the consumers after a calculated delay to increase delay precision. While we could improve upon this design and have a single manager-like thread doing all the work, currently if we have a `newSingleThreadExecutor()` and a longer delay task is submitted before a shorter one, we'd block the later and lose delay precision.

I chose Spring's `ThreadPoolExecutor` as default because it has a more robust implementation than Java's default `ThreadPoolExecutor` - for instance, by default, on `shutdown()` it will actually call `shutdownNow()` under the hood and cancel all remaining tasks, which is perfect for our use-case since there's nothing for the threads to do as cleanup.

What I think is missing here is having a default limit on the thread pool size, and perhaps making it clearer for users that this is being used, while also letting them easily change the pool size limit. Also, I could have used Java's `Executor` interface instead of Spring's `TaskExecutor` for this - not sure if it's worth changing this though since in `Spring Boot` applications users usually use the Spring interface anyway.

Even if we switched to a plain JDK `Executor` such as the ones returned by the `Executors` class, unless I'm missing something, it'd [still be necessary](https://jenkov.com/tutorials/java-util-concurrent/executorservice.html#executorservice-shutdown) to call `shutdown()` somewhere to make sure the `Executor` doesn't accept new tasks and let it free up resources. That's why in `ThreadPoolTaskExecutor` `shutdown()`  is called in the `destroy()` method - if our thread pool was a bean, we'd be able to leverage that, and thus free ourselves from doing it manually.

Reply:
I think I can come up with a `KafkaBackOffManagerConfigurer` in the `RetryTopicConfigurationSupport` class to make all these options more explicit to the user.

Reply:
The `Executors.newSingleThreadExecutor()` is a `FinalizableDelegatedExecutorService`. So, it frees its resources when the instance is destroyed by garbage collector.

But yeah... to much attention to it.
Let's leave it as is for now and we always can revise it later (if any)!

Reply:
That's nice, I wasn't aware of that. But I think your concern with using a `ThreadPoolTaskExecutor` is valid, and I think it's worth making it clear in the `Support` class that such an expensive resource is being used, while also allowing users to easily customize that. Maybe even add a small entry to the documentation about it. I'll work on these changes.

And it's a good thing to keep in mind to revise this solution when we have the time so we can use a single-threaded approach instead.

Thanks a lot for your input!